### PR TITLE
[Gluon] Add VAE demo

### PR DESCRIPTION
**Pull request by xidulu**
## Description ##
Implement VAE with gluon.probability

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by szhengac** [comment]
> Should we use https://pypi.org/project/notedown/ instead of committing the raw notebook? Would you recommend integrating the example into the website as well?

Yes. I agree that using markdown is better. It is a good example to showcase the usuage of probability module so I think we can put it into the website.

**Comment by xidulu** [comment]
> Should we use https://pypi.org/project/notedown/ instead of committing the raw notebook? Would you recommend integrating the example into the website as well?

@leezu That's sounds great, are there any existing examples that I can follow?

**Comment by xidulu** [comment]
@szha @leezu 
I am wondering if MXNet's website currently support automatically converting markdown demos to notebook via notedown, like GluonNLP: https://gluon-nlp.mxnet.io/examples/sentiment_analysis/sentiment_analysis.html ? Or I believe it could be little bit weird to provide the example in a markdown form.

**Comment by szha** [comment]
not yet. I think we should provide the automation in CI first, which may take some time. in the meantime, I think we can check in the jupyter notebook for now with the expectation that it will be replaced by notedown in the future.

**Comment by leezu** [comment]
@xidulu @szha it's supported via https://github.com/apache/incubator-mxnet/blob/243ade93bcb8b7962d1faeb89c98409e3ae0d7a4/docs/python_docs/python/Makefile#L31-L33

**Comment by xidulu** [comment]
@leezu 
Thx for pointing that out, I will take a look at it.

File: example/probability/VAE/VAE.md [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+```{.python .input}
+import numpy as np
+import mxnet as mx
+from mxnet import autograd, gluon, np, npx
+from mxnet.gluon import nn
+import mxnet.gluon.probability as mgp
+import matplotlib.pyplot as plt
+
+npx.set_np()
```
Comment (szhengac) [inline example/probability/VAE/VAE.md:32]: add comment on why we include this line. most of users are not familiar with numpy api.

File: example/probability/VAE/VAE.md [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            self.decoder.add(nn.Dense(n_hidden, activation=act_type))
+        self.decoder.add(nn.Dense(n_output, activation='sigmoid'))
+        
+    def encode(self, x):
+        r"""
+        Given a batch of x,
+        return the encoder's output
+        """
+        h = self.encoder(x)
+        loc_scale = np.split(h, 2, 1)
```
Comment (szhengac) [inline example/probability/VAE/VAE.md:89]: add comment

File: example/probability/VAE/VAE.md [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        loc = loc_scale[0]
+        log_variance = loc_scale[1]
+        scale = np.exp(0.5 * log_variance)
```
Comment (szhengac) [inline example/probability/VAE/VAE.md:92]: why do we use 0.5?

**Comment by xidulu** [comment]
@szha @leezu 

Anything else to polish?

**Comment by szha** [comment]
I'm ok with merging.

**Comment by szha** [comment]
Thanks for the demo, @xidulu!

**Pull request merged**